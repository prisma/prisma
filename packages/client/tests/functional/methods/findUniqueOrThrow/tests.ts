import { faker } from '@faker-js/faker'
import { expectTypeOf } from 'expect-type'

import testMatrix from './_matrix'
// @ts-ignore
import type { Prisma as PrismaNamespace, PrismaClient } from './node_modules/@prisma/client'

declare let prisma: PrismaClient
declare let Prisma: typeof PrismaNamespace

const existingEmail = faker.internet.email()
const nonExistingEmail = faker.internet.email()

testMatrix.setupTestSuite((_suiteConfig, _suiteMeta, clientMeta) => {
  beforeAll(async () => {
    await prisma.user.create({ data: { email: existingEmail, posts: { create: { title: 'How to exist?' } } } })
  })

  test('finds existing record', async () => {
    const record = await prisma.user.findUniqueOrThrow({ where: { email: existingEmail } })
    expect(record).toMatchObject({ id: expect.any(String), email: existingEmail })
    expectTypeOf(record).not.toBeNullable()
  })

  test('throws if record was not found', async () => {
    const record = prisma.user.findUniqueOrThrow({ where: { email: nonExistingEmail } })
    await expect(record).rejects.toThrowError(new Prisma.NotFoundError('No User found'))
  })

  // TODO: it actually does not work this way, but neither does `rejectOnNotFound`.
  // unclear, if intentional
  test.skip('works with transactions', async () => {
    const newEmail = faker.internet.email()
    const result = prisma.$transaction([
      prisma.user.create({ data: { email: newEmail } }),
      prisma.user.findUnique({ where: { email: nonExistingEmail }, rejectOnNotFound: true }),
    ])

    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`No User found`)

    const record = await prisma.user.findUnique({ where: { email: newEmail } })
    expect(record).toBeNull()
  })

  testIf(!clientMeta.dataProxy)('works with interactive transactions', async () => {
    const newEmail = faker.internet.email()
    const result = prisma.$transaction(async (prisma) => {
      await prisma.user.create({ data: { email: newEmail } })
      await prisma.user.findUniqueOrThrow({ where: { email: nonExistingEmail } })
    })

    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`No User found`)

    const record = await prisma.user.findUnique({ where: { email: newEmail } })
    expect(record).toBeNull()
  })

  test('reports correct method name in case of validation error', async () => {
    const record = prisma.user.findUniqueOrThrow({
      where: {
        // @ts-expect-error triggering validation error on purpose
        notAUserField: true,
      },
    })
    await expect(record).rejects.toMatchObject({
      message: expect.stringContaining('Invalid `prisma.user.findUniqueOrThrow()` invocation'),
    })
  })

  // TODO: Edge: skipped because of the error snapshot
  testIf(clientMeta.runtime !== 'edge')('does not accept rejectOnNotFound option', async () => {
    const record = prisma.user.findUniqueOrThrow({
      where: { email: existingEmail },
      // @ts-expect-error passing not supported option on purpose
      rejectOnNotFound: false,
    })

    await expect(record).rejects.toMatchPrismaErrorInlineSnapshot(`

      Invalid \`prisma.user.findUniqueOrThrow()\` invocation in
      /client/tests/functional/methods/findUniqueOrThrow/tests.ts:0:0

        XX 
        XX // TODO: Edge: skipped because of the error snapshot
        XX testIf(clientMeta.runtime !== 'edge')('does not accept rejectOnNotFound option', async () => {
      â†’ XX   const record = prisma.user.findUniqueOrThrow(
      'rejectOnNotFound' option is not supported
    `)
  })
})
