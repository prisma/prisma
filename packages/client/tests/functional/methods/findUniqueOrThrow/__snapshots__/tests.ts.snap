// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`methods.findUniqueOrThrow (provider=cockroachdb, runtime=edge) does not accept rejectOnNotFound option 1`] = `

Invalid \`prisma.user.findUniqueOrThrow()\` invocation:


'rejectOnNotFound' option is not supported
`;

exports[`methods.findUniqueOrThrow (provider=cockroachdb, runtime=node) does not accept rejectOnNotFound option 1`] = `

Invalid \`prisma.user.findUniqueOrThrow()\` invocation in
/client/tests/functional/methods/findUniqueOrThrow/tests.ts:0:0

  XX })
  XX 
  XX test('does not accept rejectOnNotFound option', async () => {
→ XX   const record = prisma.user.findUniqueOrThrow(
'rejectOnNotFound' option is not supported
`;

exports[`methods.findUniqueOrThrow (provider=mongodb, runtime=edge) does not accept rejectOnNotFound option 1`] = `

Invalid \`prisma.user.findUniqueOrThrow()\` invocation:


'rejectOnNotFound' option is not supported
`;

exports[`methods.findUniqueOrThrow (provider=mongodb, runtime=node) does not accept rejectOnNotFound option 1`] = `

Invalid \`prisma.user.findUniqueOrThrow()\` invocation in
/client/tests/functional/methods/findUniqueOrThrow/tests.ts:0:0

  XX })
  XX 
  XX test('does not accept rejectOnNotFound option', async () => {
→ XX   const record = prisma.user.findUniqueOrThrow(
'rejectOnNotFound' option is not supported
`;

exports[`methods.findUniqueOrThrow (provider=mysql, runtime=edge) does not accept rejectOnNotFound option 1`] = `

Invalid \`prisma.user.findUniqueOrThrow()\` invocation:


'rejectOnNotFound' option is not supported
`;

exports[`methods.findUniqueOrThrow (provider=mysql, runtime=node) does not accept rejectOnNotFound option 1`] = `

Invalid \`prisma.user.findUniqueOrThrow()\` invocation in
/client/tests/functional/methods/findUniqueOrThrow/tests.ts:0:0

  XX })
  XX 
  XX test('does not accept rejectOnNotFound option', async () => {
→ XX   const record = prisma.user.findUniqueOrThrow(
'rejectOnNotFound' option is not supported
`;

exports[`methods.findUniqueOrThrow (provider=postgresql, runtime=edge) does not accept rejectOnNotFound option 1`] = `

Invalid \`prisma.user.findUniqueOrThrow()\` invocation:


'rejectOnNotFound' option is not supported
`;

exports[`methods.findUniqueOrThrow (provider=postgresql, runtime=node) does not accept rejectOnNotFound option 1`] = `

Invalid \`prisma.user.findUniqueOrThrow()\` invocation in
/client/tests/functional/methods/findUniqueOrThrow/tests.ts:0:0

  XX })
  XX 
  XX test('does not accept rejectOnNotFound option', async () => {
→ XX   const record = prisma.user.findUniqueOrThrow(
'rejectOnNotFound' option is not supported
`;

exports[`methods.findUniqueOrThrow (provider=sqlite, runtime=node) does not accept rejectOnNotFound option 1`] = `

Invalid \`prisma.user.findUniqueOrThrow()\` invocation in
/client/tests/functional/methods/findUniqueOrThrow/tests.ts:0:0

  XX })
  XX 
  XX test('does not accept rejectOnNotFound option', async () => {
→ XX   const record = prisma.user.findUniqueOrThrow(
'rejectOnNotFound' option is not supported
`;

exports[`methods.findUniqueOrThrow (provider=sqlserver, runtime=edge) does not accept rejectOnNotFound option 1`] = `

Invalid \`prisma.user.findUniqueOrThrow()\` invocation:


'rejectOnNotFound' option is not supported
`;

exports[`methods.findUniqueOrThrow (provider=sqlserver, runtime=node) does not accept rejectOnNotFound option 1`] = `

Invalid \`prisma.user.findUniqueOrThrow()\` invocation in
/client/tests/functional/methods/findUniqueOrThrow/tests.ts:0:0

  XX })
  XX 
  XX test('does not accept rejectOnNotFound option', async () => {
→ XX   const record = prisma.user.findUniqueOrThrow(
'rejectOnNotFound' option is not supported
`;
